From fd55b3e84f2d892d927deebc81ee3dbb67e4589a Mon Sep 17 00:00:00 2001
From: Koji Ishii <kojii@chromium.org>
Date: Wed, 16 Oct 2024 22:14:23 +0000
Subject: [PATCH] [text-decoration] Fix crashes for multiline ellipsis with `::first-line`

The ellipsis should be styled according to the block[1]. It
was using the first-line style of the block, because it's the
common case, and it's the same style unless `::first-line` is
applied.

This patch changes it to distinguish ellipsis on the first
line from other cases.

Wrong style can cause wrong rendering, but when
`text-decoration` is involved, it can confuse decoration box
calculation and result in crashes. This patch fixes both the
crashes and the correctness.

[1] https://www.w3.org/TR/css-overflow-3/#ellipsing-details

Fixed: 358957652
Change-Id: I26af0ae034876bdc0b0cc114670c52d24e2f36a0
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5936973
Commit-Queue: Kent Tamura <tkent@chromium.org>
Auto-Submit: Koji Ishii <kojii@chromium.org>
Reviewed-by: Kent Tamura <tkent@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1369664}
---

--- a/third_party/blink/renderer/core/layout/inline/fragment_item.cc
+++ b/third_party/blink/renderer/core/layout/inline/fragment_item.cc
@@ -1261,8 +1261,11 @@
     case StyleVariant::kFirstLine:
       ostream << "FirstLine";
       break;
-    case StyleVariant::kEllipsis:
-      ostream << "Ellipsis";
+    case StyleVariant::kStandardEllipsis:
+      ostream << "StandardEllipsis";
+      break;
+    case StyleVariant::kFirstLineEllipsis:
+      ostream << "FirstLineEllipsis";
       break;
   }
   return ostream << "}";
--- a/third_party/blink/renderer/core/layout/inline/fragment_item.h
+++ b/third_party/blink/renderer/core/layout/inline/fragment_item.h
@@ -173,7 +173,7 @@
     return static_cast<StyleVariant>(style_variant_);
   }
   bool UsesFirstLineStyle() const {
-    return GetStyleVariant() == StyleVariant::kFirstLine;
+    return blink::UsesFirstLineStyle(GetStyleVariant());
   }
   // Returns the style for this fragment.
   //
@@ -397,9 +397,7 @@
   }
 
   // True if this is an ellpisis generated by `text-overflow: ellipsis`.
-  bool IsEllipsis() const {
-    return GetStyleVariant() == StyleVariant::kEllipsis;
-  }
+  bool IsEllipsis() const { return blink::IsEllipsis(GetStyleVariant()); }
 
   // Returns true if the text is generated (from, e.g., list marker,
   // pseudo-element, ...) instead of from a DOM text node.
--- a/third_party/blink/renderer/core/layout/inline/inline_cursor.h
+++ b/third_party/blink/renderer/core/layout/inline/inline_cursor.h
@@ -13,6 +13,7 @@
 #include "third_party/blink/renderer/core/editing/forward.h"
 #include "third_party/blink/renderer/core/layout/inline/fragment_item.h"
 #include "third_party/blink/renderer/core/layout/inline/fragment_items.h"
+#include "third_party/blink/renderer/core/layout/style_variant.h"
 #include "third_party/blink/renderer/platform/text/text_direction.h"
 #include "third_party/blink/renderer/platform/wtf/allocator/allocator.h"
 #include "third_party/blink/renderer/platform/wtf/text/string_view.h"
@@ -33,7 +34,6 @@
 class Node;
 class PhysicalBoxFragment;
 class ShapeResultView;
-enum class StyleVariant;
 struct LayoutSelectionStatus;
 struct PhysicalOffset;
 struct PhysicalRect;
@@ -107,7 +107,7 @@
   // |ComputedStyle| and related functions.
   StyleVariant GetStyleVariant() const { return item_->GetStyleVariant(); }
   bool UsesFirstLineStyle() const {
-    return GetStyleVariant() == StyleVariant::kFirstLine;
+    return blink::UsesFirstLineStyle(GetStyleVariant());
   }
   const ComputedStyle& Style() const { return item_->Style(); }
 
--- a/third_party/blink/renderer/core/layout/inline/line_truncator.cc
+++ b/third_party/blink/renderer/core/layout/inline/line_truncator.cc
@@ -35,7 +35,8 @@
 LineTruncator::LineTruncator(const LineInfo& line_info)
     : line_style_(&line_info.LineStyle()),
       available_width_(line_info.AvailableWidth() - line_info.TextIndent()),
-      line_direction_(line_info.BaseDirection()) {}
+      line_direction_(line_info.BaseDirection()),
+      use_first_line_style_(line_info.UseFirstLineStyle()) {}
 
 const ComputedStyle& LineTruncator::EllipsisStyle() const {
   // The ellipsis is styled according to the line style.
@@ -87,7 +88,9 @@
   DCHECK(ellipsis_text_);
   DCHECK(ellipsis_shape_result_);
   line_box->AddChild(
-      *ellipsized_layout_object, StyleVariant::kEllipsis,
+      *ellipsized_layout_object,
+      use_first_line_style_ ? StyleVariant::kFirstLineEllipsis
+                            : StyleVariant::kStandardEllipsis,
       ellipsis_shape_result_, ellipsis_text_,
       LogicalRect(ellipsis_inline_offset, -ellipsis_metrics.ascent,
                   ellipsis_width_, ellipsis_metrics.LineHeight()),
--- a/third_party/blink/renderer/core/layout/inline/line_truncator.h
+++ b/third_party/blink/renderer/core/layout/inline/line_truncator.h
@@ -25,7 +25,7 @@
   STACK_ALLOCATED();
 
  public:
-  LineTruncator(const LineInfo& line_info);
+  explicit LineTruncator(const LineInfo& line_info);
 
   // Truncate |line_box| and place ellipsis. Returns the new inline-size of the
   // |line_box|.
@@ -98,6 +98,8 @@
   // This data member is available between SetupEllipsis() and
   // PlaceEllipsisNextTo().
   ShapeResultView* ellipsis_shape_result_ = nullptr;
+
+  bool use_first_line_style_ = false;
 };
 
 }  // namespace blink
--- a/third_party/blink/renderer/core/layout/layout_object.cc
+++ b/third_party/blink/renderer/core/layout/layout_object.cc
@@ -2566,13 +2566,19 @@
       if (IsInline() && IsAtomicInlineLevel())
         return StyleRef();
       return FirstLineStyleRef();
-    case StyleVariant::kEllipsis:
+    case StyleVariant::kStandardEllipsis:
       // The ellipsis is styled according to the line style.
       // https://www.w3.org/TR/css-overflow-3/#ellipsing-details
-      // Use first-line style if exists since most cases it is the first line.
       DCHECK(IsInline());
-      if (LayoutObject* block = ContainingBlock())
+      if (const LayoutObject* block = ContainingBlock()) {
+        return block->StyleRef();
+      }
+      return StyleRef();
+    case StyleVariant::kFirstLineEllipsis:
+      DCHECK(IsInline());
+      if (const LayoutObject* block = ContainingBlock()) {
         return block->FirstLineStyleRef();
+      }
       return FirstLineStyleRef();
   }
   NOTREACHED_IN_MIGRATION();
--- a/third_party/blink/renderer/core/layout/physical_fragment.h
+++ b/third_party/blink/renderer/core/layout/physical_fragment.h
@@ -320,7 +320,7 @@
     return static_cast<StyleVariant>(style_variant_);
   }
   bool UsesFirstLineStyle() const {
-    return GetStyleVariant() == StyleVariant::kFirstLine;
+    return blink::UsesFirstLineStyle(GetStyleVariant());
   }
 
   // Returns the style for this fragment.
--- a/third_party/blink/renderer/core/layout/style_variant.h
+++ b/third_party/blink/renderer/core/layout/style_variant.h
@@ -5,21 +5,35 @@
 #ifndef THIRD_PARTY_BLINK_RENDERER_CORE_LAYOUT_STYLE_VARIANT_H_
 #define THIRD_PARTY_BLINK_RENDERER_CORE_LAYOUT_STYLE_VARIANT_H_
 
+#include <cstdint>
+
 #include "base/compiler_specific.h"
 
 namespace blink {
 
 // LayoutObject can have multiple style variations.
-enum class StyleVariant { kStandard, kFirstLine, kEllipsis };
+enum class StyleVariant : uint8_t {
+  kStandard = 0,
+  kFirstLine = 1,
+  kStandardEllipsis = 2,
+  kFirstLineEllipsis = 3
+};
+
+inline bool UsesFirstLineStyle(StyleVariant variant) {
+  return static_cast<uint8_t>(variant) &
+         static_cast<uint8_t>(StyleVariant::kFirstLine);
+}
+
+inline bool IsEllipsis(StyleVariant variant) {
+  return static_cast<uint8_t>(variant) &
+         static_cast<uint8_t>(StyleVariant::kStandardEllipsis);
+}
 
 inline StyleVariant ToParentStyleVariant(StyleVariant variant) {
-  // Ancestors of ellipsis should not use the ellipsis style variant. Use
-  // first-line style if exists since most cases it is the first line. See also
-  // |LayoutObject::SlowEffectiveStyle|.
-  if (variant == StyleVariant::kEllipsis) [[unlikely]] {
-    return StyleVariant::kFirstLine;
-  }
-  return variant;
+  // Ancestors of ellipsis should not use the ellipsis style variant.
+  return static_cast<StyleVariant>(
+      static_cast<uint8_t>(variant) &
+      ~static_cast<uint8_t>(StyleVariant::kStandardEllipsis));
 }
 
 }  // namespace blink
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-ui/crashtests/text-overflow-ellipsis-multiline-crash.html
@@ -0,0 +1,15 @@
+<!DOCTYPE html>
+<style>
+div {
+  width: 5ch;
+  overflow: hidden;
+  text-overflow: ellipsis;
+}
+div::first-line {
+  text-decoration: underline;
+}
+</style>
+<div>
+  <br>
+  1234567
+</div>
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-ui/reference/text-overflow-ellipsis-multiline-001-ref.html
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<style>
+div {
+  width: 6ch;
+  overflow: hidden;
+  text-overflow: ellipsis;
+}
+div.first-line {
+  color: orange;
+}
+</style>
+<div class="first-line">1234567</div>
+<div>1234567</div>
--- /dev/null
+++ b/third_party/blink/web_tests/external/wpt/css/css-ui/text-overflow-ellipsis-multiline-001.html
@@ -0,0 +1,18 @@
+<!DOCTYPE html>
+<title>Test multiline ellipsis with `::first-line`</title>
+<link rel="match" href="reference/text-overflow-ellipsis-multiline-001-ref.html">
+<link rel="help" href="https://drafts.csswg.org/css-ui-3/#ellipsing-details">
+<style>
+div {
+  width: 6ch;
+  overflow: hidden;
+  text-overflow: ellipsis;
+}
+div::first-line {
+  color: orange;
+}
+</style>
+<div>
+  1234567<br>
+  1234567
+</div>
